src/hashsum/hashsum.c

```diff
@@ -0,0 +1,547 @@
+                    cleanupCryptoAPI(_settings);
+                    exit(EXIT_FAILURE);
+                }
+            }
+
+            LPWSTR lpwCalculatedOutput = calculateFileHash(_settings, pbFileNames[j]);
+            LPCWSTR cmpResult = _wcsicmp(lpwCalculatedOutput, pbFileOutput[j]) ? L"FAILED" : L"OK";
+
+            wprintf_s(L"%s: %s\n", pbFileNames[j], cmpResult);
+
+            HeapFree(GetProcessHeap(), 0, pbFileNames[j]);
+            HeapFree(GetProcessHeap(), 0, pbFileOutput[j]);
+            HeapFree(GetProcessHeap(), 0, lpwCalculatedOutput);
+        }
+
+        HeapFree(GetProcessHeap(), 0, pbFileNames);
+        HeapFree(GetProcessHeap(), 0, pbFileOutput);
+    }
+}
+
+/*
+ * parses a hash-files for hash-filename-pairs.
+ * 
+ * skips linecomments starting with '#'.
+ * --------------------------------------------
+ * 
+ * _IN:
+ *      _fileName: the name/path of the hash-pbFile
+ * 
+ * _IN_OUT:
+ *      _pvFileNames: the parsed filenames
+ *      _pvFileHashes: the parsed hash-strings
+ * 
+ * _RETURNS: the size of the _pvFile*-vectors
+ */
+SIZE_T readHashFile(LPWSTR **_pvFileNames, LPWSTR **_pvFileHashes, LPWSTR _fileName)
+{
+    int allocCount = 2;
+    SIZE_T counter = 0;
+
+    if (! ((*_pvFileNames) = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(LPWSTR))) ||
+        ! ((*_pvFileHashes) = HeapAlloc(GetProcessHeap(), HEAP_ZERO_MEMORY, sizeof(LPWSTR))))
+    {
+        // we have to guard the calls to HeapFree() to make sure we don't call it with NULL
+        if ((*_pvFileNames)) HeapFree(GetProcessHeap(), 0, (*_pvFileNames));
+        if ((*_pvFileHashes)) HeapFree(GetProcessHeap(), 0, (*_pvFileHashes));
+
+        (*_pvFileNames) = NULL;
+        (*_pvFileHashes) = NULL;
+
+        return counter;
+    }
+    
+    FILE *pHashFile;
+    if (_wfopen_s(&pHashFile, _fileName, L"r"))
+    {
+        WCHAR errMsg[BUFSIZ];
+        _wcserror_s((WCHAR *)&errMsg, BUFSIZ, errno);
+        _fwprintf_p(stderr, L"* %s: %s\n", _fileName, errMsg);
+
+        HeapFree(GetProcessHeap(), 0, (*_pvFileNames));
+        HeapFree(GetProcessHeap(), 0, (*_pvFileHashes));
+        (*_pvFileNames) = NULL;
+        (*_pvFileHashes) = NULL;
+        
+        return counter;
+    }
+
+    WCHAR inBuff[BUFSIZ*2];
+    while (fgetws(inBuff, (BUFSIZ*2)-1, pHashFile))
+    {
+        if (inBuff[0] == '#') continue; // TODO: better handling
+        else if (isStringBlankW(inBuff, BUFSIZ*2)) continue;
+
+        LPWSTR pbContext;
+        LPWSTR pbHash;
+        LPWSTR pbFile;
+
+        if ((pbHash = wcstok_s(inBuff, L" \t", &pbContext)) &&
+            (pbFile = wcstok_s(NULL, L" \t", &pbContext)))
+        {
+            pbFile[wcslen(pbFile)-1] = '\0';
+        }
+        else
+        {
+            wprintf_s(L"* WARNING: Maleformatted Line: %zd\n", counter+1);
+            continue;
+        }
+
+        SIZE_T cFile = wcslen(pbFile) + 1;
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Does not handle strings that are not \0-terminated; if given one it may perform an over-read (it could cause a crash if unprotected) (CWE-126).  ](https://app.codacy.com/manual/HolgerDoerner/TermTools/pullRequest?prid=4762146)